---
title: "Weekend Assignment 2"
author: "YOUR NAME HERE"
date: "12/05/2017"
output:
  html_document:
    code_folding: hide
    fig_caption: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Overview

Here is a [link](https://raw.githubusercontent.com/rnabioco/eda/master/vignettes/assignment-2.Rmd) to the Rmarkdown for this assignment. Download that and update your name at the top.

For this assignment you will be generating a "metagene" plot which summarizes genomic signals over a specific genomic feature in the human genome. This type of analysis is commonly performed with Chip-Seq and Clip-Seq datasets to investigate patterns of DNA or RNA binding protein occupancy around a genomic feature (splice site, transcription start site, start/stop codons, etc.). For this assignment we will be generating a metagene profile surrounding the transcription start site that summarize the coverage of an RNA PolII Chip-Seq experiment genome-wide.

The main idea of this analysis is depicted in Figure 1. We first calculate the coverage across at each individual TSS for every gene, then we average these profiles across every gene to produce a single metagene.

```{r, echo = FALSE, fig.cap = "Figure 1. Conceptual metagene of Chip-Seq coverage over the TSS. (A) Shown are examples Chip-Seq read coverage (reads per nucleotide) across intervals +/- 1000 nucleotides surrounding the TSS for 5 genes. (B) Metagene that shows the average coverage binned in 10 nucleotide windows across a TSS metagene +/- 1000 nucleotides surrounding the TSS"}
knitr::include_graphics('https://github.com/rnabioco/eda/raw/master/vignettes/img/assignment-2-metagene.png')
```

## Files used for this analysis:

You will use the following files in this analysis:

Filename | Description
--------------------------------- | -------------------------
hg19_genes.bed.gz  | bed file with 6 fields with the start and end coordinates for every human gene (`hg19`)
hg19_genome.txt.gz | genome file containing the lengths of every human chromosome
polra2_hela_1e6.bg.gz | bedgraph file of RNA Pol II Chip-Seq coverage data (subsampled to 1 million rows), and hosted on a server accessible by URL

These files can be loaded into R by following these steps:

```{r, genes, message = F}
library(valr)
library(eda)
library(tidyverse)

genes_file <- system.file("extdata", "hg19_genes.bed.gz", package = "eda")
genes <- read_bed(genes_file, n_fields = 6)
genes
```

```{r genome}
genome_file <- system.file("extdata", "hg19_genome.txt.gz", package = "eda")
genome <- read_genome(genome_file)
genome
```

```{r bedgraph}
bed <- read_bed("https://de.cyverse.org/anon-files/iplant/home/kriemo/data/class/polra2_hela_1e6.bg.gz", 
                n_fields = 4)
# rename name to value
bg <- select(bed, everything(), value = name)

# change value from a character to a numeric value
bg <- mutate(bg, 
              value = as.numeric(value))
bg
```

## Assignment
 
 I've provides some pseudo-code to get you started with this task. The `...` is pseudo-code that you will need to fill in with your own code.
 
### Exercise 1
Generate a tibble that contains the positions of transcription start sites for all genes on the plus strand. 

```{r answer_1, eval = F}

tss <- genes %>% 
  filter(strand == "+") %>% 
  ...

tss
```

Your `tss` bed tibble that you generate should look like this:
``` r
#> # A tibble: 28,035 x 6
#>    chrom  start    end         name score strand
#>    <chr>  <int>  <dbl>        <chr> <chr>  <chr>
#>  1  chr1  11868  11869 LOC102725121     3      +
#>  2  chr1  11873  11874      DDX11L1     3      +
#>  3  chr1  30365  30366    MIR1302-2     1      +
#>  4  chr1  30365  30366    MIR1302-9     1      +
#>  5  chr1  30365  30366   MIR1302-10     1      +
#>  6  chr1  30365  30366   MIR1302-11     1      +
#>  7  chr1  69090  69091        OR4F5     1      +
#>  8  chr1 323891 323892 LOC100132287     3      +
#>  9  chr1 323891 323892 LOC100132062     3      +
#> 10  chr1 323891 323892 LOC100133331     4      +
#> # ... with 28,025 more rows
```

### Exercise 2

Generate a tibble that contains the intervals +/- 1000 nucleotides surrounding the TSS position. Conceptually, these are the intervals that are shown in Figure 1A. `valr` provides the [`bed_slop()`](https://rnabioco.github.io/valr/reference/bed_slop.html) function, which adds nucleotides to intervals. 

[`bed_slop()`](https://rnabioco.github.io/valr/reference/bed_slop.html) requires the `genome` tibble which contains the length of each chromosome. In your Rmarkdown document, speculate as to why this function might require the lengths of each chromosome to produce the correct output. 

```{r answer_2, eval = F}
slop_bed <- bed_slop(tss, genome, ...) 

slop_bed
```


Your `slop_bed` bed tibble that you generate should look like this:
``` r
#> # A tibble: 28,034 x 6
#>    chrom  start    end         name score strand
#>    <chr>  <dbl>  <dbl>        <chr> <chr>  <chr>
#>  1  chr1  10868  12869 LOC102725121     3      +
#>  2  chr1  10873  12874      DDX11L1     3      +
#>  3  chr1  29365  31366    MIR1302-2     1      +
#>  4  chr1  29365  31366    MIR1302-9     1      +
#>  5  chr1  29365  31366   MIR1302-10     1      +
#>  6  chr1  29365  31366   MIR1302-11     1      +
#>  7  chr1  68090  70091        OR4F5     1      +
#>  8  chr1 322891 324892 LOC100132287     3      +
#>  9  chr1 322891 324892 LOC100132062     3      +
#> 10  chr1 322891 324892 LOC100133331     4      +
#> # ... with 28,024 more rows
```

### Exercise 3

Now, we have obtained intervals surrounding each gene's TSS, similar to the intervals shown in Figure 1A. For our next task we are going to break up each gene's TSS interval into smaller bins of 10 nucleotides to give us a high resolution metagene. In following steps, coverage values will be averaged across all genes at the same bin. For example bin 1 represents the first 10 nucleotides of our TSS intervals (i.e. positions -1000 to -990 for every gene). 

Smaller bin values will increase the resolution of the final plot, but dramatically increase the size of your bed interval tibble (i.e. if you break every 2001 nt region into 1 nt, then for each gene you generate 2001 new intervals). Similarly, larger bin values will decrease resolution of the metagene, but compute faster.  

`valr` provides the [`bed_makewindows()`](https://rnabioco.github.io/valr/reference/bed_makewindows.htmll) function, which will split up each interval into sub-intervals (i.e windows). This function has many options to control the size of the windows generated. Read the documentation for `bed_makewindows()` to pick an option that will generate non-overlapping 10 nucleotide windows across each TSS interval. 

Note that `bed_makewindows()` will dramatically increase the size of your output tibble. **Do not select a window size less than 10, as this will generate a tibble with potentially 10s of millions of rows(!).** 

```{r answer_3, eval = F}
tss_windows <- slop_bed %>% 
  bed_makewindow(genome, ...)
```

Your output `tss_windows` tibble should look like this:
``` r
#> # A tibble: 5,634,834 x 7
#>    chrom start   end         name score strand .win_id
#>    <chr> <int> <int>        <chr> <chr>  <chr>   <int>
#>  1  chr1 10868 10878 LOC102725121     3      +       1
#>  2  chr1 10878 10888 LOC102725121     3      +       2
#>  3  chr1 10888 10898 LOC102725121     3      +       3
#>  4  chr1 10898 10908 LOC102725121     3      +       4
#>  5  chr1 10908 10918 LOC102725121     3      +       5
#>  6  chr1 10918 10928 LOC102725121     3      +       6
#>  7  chr1 10928 10938 LOC102725121     3      +       7
#>  8  chr1 10938 10948 LOC102725121     3      +       8
#>  9  chr1 10948 10958 LOC102725121     3      +       9
#> 10  chr1 10958 10968 LOC102725121     3      +      10
#> # ... with 5,634,824 more rows
```

### Exercise 4
Examine the `tss_windows` tibble that you generated in the previous exercise. What information does the `.win_id` column contain ?

### Exercise 5

We have now obtained a proper set set of intervals to calculate Chip-Seq coverage values. The Chip-seq data that we will use is derived from an Encode experiment in which RNA polII was Chip'ed in Hela-S3 cells. The Chip experiment is described [here](https://www.encodeproject.org/experiments/ENCSR000EZL/). The coverage data is supplied as a [`bigWig`](http://genome.ucsc.edu/goldenPath/help/bigWig.html) file, which is an indexed binary format, not suitable for R. I've converted the `bigWig` to a `bedGraph` file using the `bigWigToBedgraph` C program available from `USCS` (only available for [macOS](http://hgdownload.soe.ucsc.edu/admin/exe/macOSX.x86_64/) and [linux](http://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64/)). `bedGraph` files can be read into R using the `read_bed()` function. 

The `bedGraph` file contains four columns, `chrom`, `start`, `end`, and `score`. The `score` column contains the number of reads overlapping the interval. The score is normalized by the total number of reads in the experiment, and is represented as **R**eads **P**er **M**illon:

$$\frac{reads\ overlapping\ the\ interval* 1e6}{total\ number\ of\ reads}$$

```{r}
bed <- read_bed("http://amc-sandbox.ucdenver.edu/User33/tmp/polra2_hela_1e6.bg.gz", 
                n_fields = 4)
# rename name to value
bg <- select(bed, everything(), value = name)

# change value from a character to a numeric value
bg <- mutate(bg, 
              value = as.numeric(value))
bg
```

Now we would like to perform two steps:  
1) For each TSS interval, find all intervals in the `bedGraph` that overlap  
2) For each TSS interval, sum all of the `values` for each `bedGraph` interval that overlaps the TSS interval  

The [`bed_map()`](https://rnabioco.github.io/valr/reference/bed_map.html) function accomplishes both of these steps. Shown below is a schematic which illustrates the behavior of `bed_map()`.

```{r map_glyph, echo = F, fig.cap = "Imagine that each X interval is a TSS interval and the Y intervals are the Pol2 ChipSeq bedgraph intervals. The numbers 12, 20, 30, etc. indicate the  Chip-Seq coverage across each y interval. What bed_map() will do is find all intervals in y that overlap each x interval, then allow you to compute a summary statistic, in this case a sum. Conceptually this is similar to the group_by() and summarize() functions in dplyr, but instead of grouping by columns we are grouping y intervals that overlap the x intervals. If there are no Y intervals overlaping an X interval then an NA value is reported."}

x <- trbl_interval(
  ~chrom, ~start, ~end,
  'chr1',      1,      100,
  'chr1',      150,    175,
  'chr1',      200,    300
)

y <- trbl_interval(
  ~chrom, ~start, ~end, ~value,
  'chr1', 1,      20,   10,
  'chr1', 30,     50,   20,
  'chr1', 80,     95,  30,
  'chr1', 210,    230,   5,
  'chr1', 265,    290,   10
)

bed_glyph(bed_map(x, y, value = sum(value)), label = 'value')
```

You can pass functions to `bed_map()` similar to `summarize()` such as `mean`, `sum`, `var`, and name their outputs. The function outputs then become new columns in the output tibble. 

For example, imagine we have an x and y bed tibble.

```{r}
x
y
bed_map(x, y, 
        read_total = sum(value))

bed_map(x, y, 
        read_mean = mean(value))

bed_map(x, y, 
        read_mean = mean(value),
        read_total = sum(value))
```

Using `bed_map()` calculate the total coverage (i.e. `sum()`) of the Chip-Seq coverage values (the `bedGraph` named `bg`) across the tss window intervals (`tss_windows`). This function might take 30-60 seconds to complete. 

If you are having trouble getting the proper output, it may be useful to generate a smaller tss_windows bed file to test your function before running on the whole dataset (use `sample_n()`), so that you don't have to wait 30-60 second per try.  

```{r answer_5, eval = F}
chip_coverage <- bed_map(tss_windows, 
                         bg, 
                         ...)
```

The expected output is shown below. Note that the `total_reads` column name may another name depending on what you assign your summary statistic to (i.e. `total_reads = sum(value)`, or `total_coverage = sum(value)`). Also note that there are many NA values in the total_reads column. This is expected as not every gene is being transcribed in the genome, so some regions have no coverage in the ChIP-Seq experiment. 

``` r
#> # A tibble: 5,626,622 x 8
#>    chrom start   end         name score strand .win_id total_reads
#>    <chr> <int> <int>        <chr> <chr>  <chr>   <int>       <dbl>
#>  1  chr1 10868 10878 LOC102725121     3      +       1          NA
#>  2  chr1 10873 10883      DDX11L1     3      +       1          NA
#>  3  chr1 10878 10888 LOC102725121     3      +       2          NA
#>  4  chr1 10883 10893      DDX11L1     3      +       2          NA
#>  5  chr1 10888 10898 LOC102725121     3      +       3          NA
#>  6  chr1 10893 10903      DDX11L1     3      +       3          NA
#>  7  chr1 10898 10908 LOC102725121     3      +       4          NA
#>  8  chr1 10903 10913      DDX11L1     3      +       4          NA
#>  9  chr1 10908 10918 LOC102725121     3      +       5          NA
#> 10  chr1 10913 10923      DDX11L1     3      +       5          NA
#> # ... with 5,626,612 more rows
```

### Exercise 6

Now we have arrived at the exciting part of the analysis. To review, we first defined a set of intervals surrounding each TSS. Then we split these intervals into bins of sub-intervals. Finally, we summed up the Chip-Seq coverage across each sub-interval. Now our bed file `chip_coverage` contains the RNA polII coverage surrounding the TSS for every gene across the genome. Now in this last step we will summarize the coverage in each bin and plot as a metagene.

```{r answer_6, eval = F}
group_by(chip_coverage, ...) %>% 
  summarize(...) %>% 
  ggplot()
```

The x-axis values displayed will not be ideal. To center the x-axis values around zero use the following code.

Add informative `x-axis`, `y-axis`, and `main` plot titles and explain what the plot demonstrates. Additionally add a figure caption using the `fig.cap="your caption"` chunk option. 

```{r, eval = F}
region_size <- 1000
win_size <- 10

x_labels <- seq(-region_size, region_size, by = win_size * 25)
x_breaks <- seq(1, 201, by = 25)

# add this to your ggplot call:
scale_x_continuous(labels = x_labels, breaks = x_breaks) 
```

```{r, echo = FALSE, fig.cap = "Figure 2. Example metagene plot"}
knitr::include_graphics('https://github.com/rnabioco/eda/raw/master/vignettes/img/assignment-2-metagene_plot.png')
```
