<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="pbda">
<title>Class 10: Jumping into more complex R, tips, resources • pbda</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><link href="../deps/_Source%20Sans%20Pro-0.4.1/font.css" rel="stylesheet">
<link href="../deps/_Source%20Code%20Pro-0.4.1/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Class 10: Jumping into more complex R, tips, resources">
<meta property="og:description" content="pbda">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="rnabioco.github.io" src="https://plausible.io/js/plausible.js"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">pbda</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../index.html">Home</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-class-material">Class Material</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-class-material">
    <a class="dropdown-item" href="../articles/class-1.html">Class 1: Overview, Data Import and Tidying</a>
    <a class="dropdown-item" href="../articles/class-2.html">Class 2: Analyzing qPCR data</a>
    <a class="dropdown-item" href="../articles/class-5.html">Class 5: Basic Data Types and Structures</a>
    <a class="dropdown-item" href="../articles/class-6.html">Class 6: Writing Functions</a>
    <a class="dropdown-item" href="../articles/class-7.html">Class 7: Iteration</a>
    <a class="dropdown-item" href="../articles/class-8.html">Class 8: Heatmaps and Clustering</a>
    <a class="dropdown-item" href="../articles/class-9.html">Class 9: Exploring data with PCA</a>
    <a class="dropdown-item" href="../articles/class-10.html">Class 10: Jumping into more complex R, tips, resources</a>
    <a class="dropdown-item" href="../articles/class-11.html">Class 11: RNA-seq Analysis</a>
    <a class="dropdown-item" href="../articles/class-12.html">Class 12: Working with gene lists</a>
    <div class="dropdown-divider"></div>
    <a class="dropdown-item" href="../articles/assignments.html">Assignment Submission and Grading Rubric</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/rnabioco/practical-data-analysis/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Class 10: Jumping into more complex R, tips, resources</h1>
                        <h4 data-toc-skip class="author">Rui Fu</h4>
            
            <h4 data-toc-skip class="date">2022-07-06</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/rnabioco/practical-data-analysis/blob/HEAD/vignettes/class-10.Rmd" class="external-link"><code>vignettes/class-10.Rmd</code></a></small>
      <div class="d-none name"><code>class-10.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="please-fill-out-end-of-class-survey-to-improve-future-classes">Please fill out end-of-class survey to improve future classes<a class="anchor" aria-label="anchor" href="#please-fill-out-end-of-class-survey-to-improve-future-classes"></a>
</h2>
</div>
<div class="section level2">
<h2 id="class-material-can-be-accessed-at">Class material can be accessed at:<a class="anchor" aria-label="anchor" href="#class-material-can-be-accessed-at"></a>
</h2>
<p><a href="https://rnabioco.github.io/practical-data-analysis/" class="external-link uri">https://rnabioco.github.io/practical-data-analysis/</a></p>
</div>
<div class="section level2">
<h2 id="speeding-up-repeated-rmd-knitting">1. Speeding up repeated Rmd knitting<a class="anchor" aria-label="anchor" href="#speeding-up-repeated-rmd-knitting"></a>
</h2>
<p>Read the <a href="https://bookdown.org/yihui/rmarkdown/" class="external-link">Guide to RMarkdown</a> for an exhaustive description of the various formats and options for using RMarkdown documents. Note that HTMLs and slides for this class were all made from Rmd.</p>
<div class="section level3">
<h3 id="caching">Caching<a class="anchor" aria-label="anchor" href="#caching"></a>
</h3>
<p>You can speed up knitting of your Rmds by using caching to store the results from each chunk, instead of rerunning them each time. Note that if you modify the code chunk, previous caching is ignored.</p>
<p>For each chunk, set <code>{r, cache = TRUE}</code></p>
<p>Or the option can be set globally at top of the document. Like this:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/knitr/man/opts_chunk.html" class="external-link">opts_chunk</a></span><span class="op">$</span><span class="fu">set</span><span class="op">(</span>cache <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="alternatively-save-and-load--rds-rda-files">Alternatively, save and load .Rds/.Rda files<a class="anchor" aria-label="anchor" href="#alternatively-save-and-load--rds-rda-files"></a>
</h3>
<p>Run once, save, and load instead of rerunning resource intensive parts.</p>
<p>If you have non-deterministic functions, such as <code>kmeans</code>, remember to <code>set.seed</code>, or save and load result objects.</p>
</div>
</div>
<div class="section level2">
<h2 id="finding-useful-packages">2. Finding useful packages<a class="anchor" aria-label="anchor" href="#finding-useful-packages"></a>
</h2>
<p>In most cases, what you need is already made into well-documented packages, and you don’t have to reinvent the wheel (but sometimes you should?). Depending on where the package is curated, installation is different. Some examples below:</p>
<ol style="list-style-type: decimal">
<li>
<code>Gviz</code> - visualize gene model</li>
<li>
<code>VennDiagram</code> - making custom venn diagrams</li>
<li>
<code>emo</code> - inserting emojis into Rmd</li>
</ol>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># BiocManager::install("Gviz") # from bioconductor</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/vignette.html" class="external-link">vignette</a></span><span class="op">(</span><span class="st">"Gviz"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># install.packages("eulerr") # from CRAN</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu">eulerr</span><span class="fu">::</span><span class="fu">euler</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>set1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"geneA"</span>, <span class="st">"geneB"</span>, <span class="st">"geneC"</span><span class="op">)</span>, </span>
<span>                       set2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"geneC"</span>, <span class="st">"geneD"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># devtools::install_github("hadley/emo") # from github</span></span>
<span><span class="fu">emo</span><span class="fu">::</span><span class="fu">ji</span><span class="op">(</span><span class="st">"smile"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="git---version-control">3. Git - version control<a class="anchor" aria-label="anchor" href="#git---version-control"></a>
</h2>
<p>Advantages of using version control include:</p>
<ol style="list-style-type: decimal">
<li><p>rolling back code if needed</p></li>
<li><p>branched development, tackling individual issues/tasks</p></li>
<li><p>collaboration, etc</p></li>
</ol>
<p>Git was first created by Linus Torvalds for coordinating development of Linux. Read this guide for <a href="https://git-scm.com/book/en/v2/Getting-Started-About-Version-Control" class="external-link">Getting started</a> and try out the <a href="https://learngitbranching.js.org/" class="external-link">Interactive practice</a>.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># for bioinformatics, get comfortable with command line too</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="co"># and avoid using windows for compatibility issues</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="fu">git</span> status <span class="co"># list changes</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="fu">git</span> blame assignments.Rmd <span class="co"># see who contributed</span></span></code></pre></div>
<p>This can be handled by Rstudio as well (new tab next to <code>Connections</code> and <code>Build</code>)</p>
<div class="section level3">
<h3 id="put-your-code-on-github">Put your code on GitHub<a class="anchor" aria-label="anchor" href="#put-your-code-on-github"></a>
</h3>
<p>As you write more code, especially as functions and script pipelines, hosting and documenting them on GitHub is great way to make them portable and searchable. Even the free tier of GitHub accounts now has private repositories (repo).</p>
<p>If you have any interest in a career in data science/informatics, GitHub is also a common showcase of what (and how well/often) you can code. After some accumulation of code, definitely put your GitHub link on your CV/resume.</p>
</div>
<div class="section level3">
<h3 id="example-repo-rbi---valr">Example repo (RBI) - <a href="https://github.com/rnabioco/valr" class="external-link">valr</a><a class="anchor" aria-label="anchor" href="#example-repo-rbi---valr"></a>
</h3>
<p>Refer to package examples and <a href="https://r-pkgs.org/index.html" class="external-link">guide to building packages</a> to build your own package and website documentation.</p>
</div>
<div class="section level3">
<h3 id="asking-for-help-with-other-packages-on-github">Asking for help with other packages on GitHub<a class="anchor" aria-label="anchor" href="#asking-for-help-with-other-packages-on-github"></a>
</h3>
<p>Every package should include README, installation instructions, and a maintained <code>issues</code> page where questions and bugs can be reported and addressed. Example: <a href="https://github.com/tidyverse/readr/issues" class="external-link">readr GitHub page</a> Don’t be afraid to file new issues, but very often your problems are already answered in the <code>closed</code> section.</p>
</div>
</div>
<div class="section level2">
<h2 id="other-tipsbest-practices">4. Other tips/best practices<a class="anchor" aria-label="anchor" href="#other-tipsbest-practices"></a>
</h2>
<div class="section level3">
<h3 id="a--code-and-data-organization">a. Code and data organization<a class="anchor" aria-label="anchor" href="#a--code-and-data-organization"></a>
</h3>
<p>Read this: <a href="https://doi.org/10.1371/journal.pcbi.1000424" class="external-link">A Quick Guide to Organizing Computational Biology Projects</a></p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="ex">NEWS.md</span>             <span class="co"># markdown document for tracking progress</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="ex">data</span>                <span class="co"># data directory for storing raw and processed data</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="ex">dbases</span>              <span class="co"># database files downloaded from other places</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="ex">docs</span>                <span class="co"># publication documents and random project files</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="ex">your-project.Rproj</span>  <span class="co"># Make an Rproject in Rstudio for every project</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="ex">results</span>             <span class="co"># store all RMarkdown here</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="ex">src</span>                 <span class="co"># store useful scripts here</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="a2--here-folder-structure-management">a2. <code>here</code>, folder structure management<a class="anchor" aria-label="anchor" href="#a2--here-folder-structure-management"></a>
</h3>
<p><a href="https://github.com/jennybc/here_here" class="external-link uri">https://github.com/jennybc/here_here</a></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html" class="external-link">here</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># always points to top-level of current project</span></span>
<span></span>
<span><span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html" class="external-link">here</a></span><span class="op">(</span><span class="st">"vignettes"</span>, <span class="st">"class-12.Rmd"</span><span class="op">)</span> <span class="co"># never confused about folder structure</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="b--styler-clean-up-code-readability">b. <code>styler</code>, clean up code readability<a class="anchor" aria-label="anchor" href="#b--styler-clean-up-code-readability"></a>
</h3>
<p>Refer to this <a href="http://adv-r.had.co.nz/Style.html" class="external-link">style guide</a> often, so you don’t have to go back to make the code readable/publishable later.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">styler</span><span class="fu">::</span><span class="fu">style_text</span><span class="op">(</span><span class="st">"</span></span>
<span><span class="st">my_fun &lt;- function(x, </span></span>
<span><span class="st">y, </span></span>
<span><span class="st">z) {</span></span>
<span><span class="st">  x+</span></span>
<span><span class="st">    z</span></span>
<span><span class="st">}</span></span>
<span><span class="st">                   "</span><span class="op">)</span></span>
<span><span class="co">#&gt; my_fun &lt;- function(x,</span></span>
<span><span class="co">#&gt;                    y,</span></span>
<span><span class="co">#&gt;                    z) {</span></span>
<span><span class="co">#&gt;   x +</span></span>
<span><span class="co">#&gt;     z</span></span>
<span><span class="co">#&gt; }</span></span>
<span></span>
<span><span class="co"># styler::style_file   # for an entire file</span></span>
<span><span class="co"># styler::style_dir    # or folder</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="c--benchmarking-with-microbenchmark-and-profvis">c. Benchmarking, with <code>microbenchmark</code> and <code>profvis</code><a class="anchor" aria-label="anchor" href="#c--benchmarking-with-microbenchmark-and-profvis"></a>
</h3>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># example, compare base and readr csv reading functions</span></span>
<span><span class="va">path_to_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"gene_tibble.csv"</span>, package <span class="op">=</span> <span class="st">"pbda"</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">microbenchmark</span><span class="fu">::</span><span class="fu">microbenchmark</span><span class="op">(</span></span>
<span>  base <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="va">path_to_file</span><span class="op">)</span>,</span>
<span>  readr <span class="op">=</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="va">path_to_file</span><span class="op">)</span>,</span>
<span>  times <span class="op">=</span> <span class="fl">5</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">res</span>, signif <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu">microbenchmark</span><span class="fu">:::</span><span class="fu">autoplot.microbenchmark</span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># example, looking at each step of a script</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">profvis</span><span class="fu">::</span><span class="fu">profvis</span><span class="op">(</span><span class="op">{</span></span>
<span>  <span class="va">path_to_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"gene_tibble.csv"</span>, package <span class="op">=</span> <span class="st">"pbda"</span><span class="op">)</span></span>
<span>  <span class="va">c1</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="va">path_to_file</span><span class="op">)</span></span>
<span>  <span class="va">c1</span> <span class="op">&lt;-</span> <span class="va">c1</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">0.00001</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>gene <span class="op">=</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/case.html" class="external-link">str_to_lower</a></span><span class="op">(</span><span class="va">gene</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="va">p</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="d--shiny-interactive-web-app-for-data-exploration">d. <code>shiny</code>, interactive web app for data exploration<a class="anchor" aria-label="anchor" href="#d--shiny-interactive-web-app-for-data-exploration"></a>
</h3>
<p>Making an interactive interface to data and plotting is easy in R. Examples and corresponding code can be found at <a href="https://shiny.rstudio.com/gallery/" class="external-link uri">https://shiny.rstudio.com/gallery/</a>.</p>
</div>
<div class="section level3">
<h3 id="e--running-r-script-from-terminal">e. Running R script from terminal<a class="anchor" aria-label="anchor" href="#e--running-r-script-from-terminal"></a>
</h3>
<p>Save functions in .R files to easily <code><a href="https://rdrr.io/r/base/source.html" class="external-link">source()</a></code> them later. If you have code that does not need input from beginning to end, .R files can be executed from the command line/terminal as well.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># example, mat.R that makes matrix and does calculations</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="ex">Rscript</span> /Users/rf/teach/practical-data-analysis/R/mat.R</span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="finding-help-online">5. Finding help online<a class="anchor" aria-label="anchor" href="#finding-help-online"></a>
</h2>
<p><img src="https://imgs.xkcd.com/comics/wisdom_of_the_ancients.png"></p>
<p>The <a href="https://community.rstudio.com/" class="external-link">R studio community forums</a> are a great resource for asking questions about tidyverse related packages.</p>
<p><a href="https://stackoverflow.com/" class="external-link">StackOverflow</a> provides user-contributed questions and answers on a variety of topics.</p>
<p>For help with bioconductor packages, visit the <a href="https://support.bioconductor.org/" class="external-link">Bioc support page</a></p>
<p>Find out if others are having similar issues by searching the issue on the package <a href="https://github.com/" class="external-link">GitHub</a> page.</p>
<div class="section level3">
<h3 id="general-r-coding-resources">General R coding resources<a class="anchor" aria-label="anchor" href="#general-r-coding-resources"></a>
</h3>
<p>Hadley Wickham has written very good (free) ebooks on using R.</p>
<p><a href="https://r4ds.had.co.nz/" class="external-link">R for Data Science</a></p>
<p><a href="http://adv-r.had.co.nz/" class="external-link">Advanced R</a></p>
</div>
<div class="section level3">
<h3 id="bioinformatic-resources">Bioinformatic resources<a class="anchor" aria-label="anchor" href="#bioinformatic-resources"></a>
</h3>
<p>For general bioinformatics advice, we found the following text very useful.</p>
<p><a href="http://vincebuffalo.org/book/" class="external-link">Bioinformatics Data Skills</a> by Vince Buffalo</p>
<p>For statistics related to bioinformatics, this free course is excellent:</p>
<p><a href="http://genomicsclass.github.io/book/" class="external-link">PH525x series - Biomedical Data Science</a></p>
<p>For more detailed descriptions of single cell RNA-Seq analysis:</p>
<p><a href="https://scrnaseq-course.cog.sanger.ac.uk/website/index.html" class="external-link uri">https://scrnaseq-course.cog.sanger.ac.uk/website/index.html</a></p>
<p><a href="https://satijalab.org/seurat/vignettes.html" class="external-link uri">https://satijalab.org/seurat/vignettes.html</a></p>
</div>
<div class="section level3">
<h3 id="cheat-sheets">Cheat sheets<a class="anchor" aria-label="anchor" href="#cheat-sheets"></a>
</h3>
<p>Rstudio links to common ones here: <code>Help</code> -&gt; <code>Cheatsheets</code>. More are hosted online, such as for <a href="https://rstudio.com/wp-content/uploads/2016/09/RegExCheatsheet.pdf" class="external-link">regular expressions</a>.</p>
<p>Useful to keep your own stash too.</p>
</div>
</div>
<div class="section level2">
<h2 id="offline-help">6. Offline help<a class="anchor" aria-label="anchor" href="#offline-help"></a>
</h2>
<p>The RBI fellows hold standing office hours on Thursdays, walk-ins 1-2pm, by appointment 2-4pm, in RC1-South Room 9101. We are happy to help out with coding and RNA/DNA-related informatics questions. Send us an email to let us know you will be stopping by (<code>rbi.fellows@cuanschutz.edu</code>).</p>
</div>
<div class="section level2">
<h2 id="sometimes-code-is-just-broken">7. Sometimes code is just broken<a class="anchor" aria-label="anchor" href="#sometimes-code-is-just-broken"></a>
</h2>
<p>No one writes perfect code. A certain highly-regarded and cited scRNA-seq package drew cell type progression arrows in the wrong direction, for months before the issue was fixed!</p>
<p>If you suspect bugs or mishandled edge cases, go to the package GitHub and search the issues section to see if the problem has been reported or fixed. If not, submit an issue that describes the problem. The <a href="https://reprex.tidyverse.org/" class="external-link">reprex package</a> makes it easy to produce well-formatted reproducible examples that demonstrate the problem. Often developers will be thankful for your help with making their software better.</p>
</div>
<div class="section level2">
<h2 id="use-r-often--dont-use-excel">8. Use R often. Don’t use Excel!<a class="anchor" aria-label="anchor" href="#use-r-often--dont-use-excel"></a>
</h2>
<p>Excel:</p>
<ol style="list-style-type: decimal">
<li><p>Gets unwieldy with large files</p></li>
<li><p>Poor plotting options</p></li>
<li><p>Hard to do dplyr-like manipulations</p></li>
<li><p>Hard to integrate with other informatics tools</p></li>
<li><p>Annoying to apply same analysis over multiple files</p></li>
<li><p>And random things like this:</p></li>
</ol>
<div class="figure">
<img src="https://raw.githubusercontent.com/IDPT7810/practical-data-analysis/master/inst/extdata/noexcel.png" style="width:70.0%" alt=""><p class="caption"><a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-1044-7" class="external-link uri">https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-1044-7</a></p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>An R package developed by the <a href="https://medschool.cuanschutz.edu/rbi" class="external-link">RNA Bioscience Initiative</a></p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
