<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="pbda">
<title>Problem Set 4 • pbda</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><link href="../deps/_Source%20Sans%20Pro-0.4.1/font.css" rel="stylesheet">
<link href="../deps/_Source%20Code%20Pro-0.4.1/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Problem Set 4">
<meta property="og:description" content="pbda">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="rnabioco.github.io" src="https://plausible.io/js/plausible.js"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">pbda</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../index.html">Home</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-class-material">Class Material</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-class-material">
    <a class="dropdown-item" href="../articles/class-1.html">Class 1: Overview, Data Import and Tidying</a>
    <a class="dropdown-item" href="../articles/class-2.html">Class 2: Analyzing qPCR data</a>
    <a class="dropdown-item" href="../articles/class-5.html">Class 5: Basic Data Types and Structures</a>
    <a class="dropdown-item" href="../articles/class-6.html">Class 6: Writing Functions</a>
    <a class="dropdown-item" href="../articles/class-7.html">Class 7: Iteration</a>
    <a class="dropdown-item" href="../articles/class-8.html">Class 8: Heatmaps and Clustering</a>
    <a class="dropdown-item" href="../articles/class-9.html">Class 9: Exploring data with PCA</a>
    <a class="dropdown-item" href="../articles/class-10.html">Class 10: Jumping into more complex R, tips, resources</a>
    <a class="dropdown-item" href="../articles/class-11.html">Class 11: RNA-seq Analysis</a>
    <a class="dropdown-item" href="../articles/class-12.html">Class 12: Working with gene lists</a>
    <div class="dropdown-divider"></div>
    <a class="dropdown-item" href="../articles/assignments.html">Assignment Submission and Grading Rubric</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/rnabioco/practical-data-analysis/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Problem Set 4</h1>
                        <h4 data-toc-skip class="author">Your Name Here</h4>
            
            <h4 data-toc-skip class="date">12/11/2018</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/rnabioco/practical-data-analysis/blob/HEAD/vignettes/problem-set-4.Rmd" class="external-link"><code>vignettes/problem-set-4.Rmd</code></a></small>
      <div class="d-none name"><code>problem-set-4.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>Here is a <a href="https://raw.githubusercontent.com/IDPT7810/practical-data-analysis/master/vignettes/problem-set-4.Rmd" class="external-link">link</a> to the text of these exercises.</p>
</div>
<div class="section level2">
<h2 id="question-1">Question 1<a class="anchor" aria-label="anchor" href="#question-1"></a>
</h2>
<p>A second 10X Genomics single cell RNA-seq dataset is included in the pbda package. This dataset is derived from a human solid tumor. Get its path using the code below:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tumor_10x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"tumor_gene_bc_matrices"</span>, <span class="st">"hg19"</span>, package <span class="op">=</span> <span class="st">"pbda"</span><span class="op">)</span></span></code></pre></div>
<p>Using the new data, create a new Seurat object named <code>tumor</code>. Next, extract the list of mitochondrial genes in the matrix, calculate the percentage of reads that are derived from the mitochondrial genome in each cell, and add this to a metadata slot called “percent.mito” in the <code>tumor</code> object.</p>
<p>Now, using the <code>GenePlot()</code> function, plot nUMI vs. nGene and nUMI vs. percent.mito. Based on these plots, filter the data to remove dead cells and likely doublets (using <code>FilterCells()</code>).</p>
<div class="section level3">
<h3 id="strategy">Strategy<a class="anchor" aria-label="anchor" href="#strategy"></a>
</h3>
</div>
<div class="section level3">
<h3 id="interpretation">Interpretation<a class="anchor" aria-label="anchor" href="#interpretation"></a>
</h3>
</div>
</div>
<div class="section level2">
<h2 id="question-2">Question 2<a class="anchor" aria-label="anchor" href="#question-2"></a>
</h2>
<p>Now that you’ve filtered your cells, normalize the data (<code>NormalizeData()</code>) then identify variable genes using the code below (the cutoffs have been adjusted for this dataset - don’t worry about interpreting the VMR plot):</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tumor</span> <span class="op">&lt;-</span> <span class="fu">FindVariableGenes</span><span class="op">(</span><span class="va">tumor</span>,</span>
<span>                          mean.function <span class="op">=</span> <span class="va">ExpMean</span>,</span>
<span>                          dispersion.function <span class="op">=</span> <span class="va">LogVMR</span>, </span>
<span>                          x.low.cutoff <span class="op">=</span> <span class="fl">0.025</span>,</span>
<span>                          x.high.cutoff <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                          y.cutoff <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p>Next, scale the data and run PCA on your <code>tumor</code> object using the variable genes you’ve just identified.</p>
<p>Finally, use an elbow plot and/or PC heatmaps to select the prinicpal components to use for tSNE and clustering. Justify your choice of PCs (do they capture the full variation present in the data?).</p>
<div class="section level3">
<h3 id="strategy-1">Strategy<a class="anchor" aria-label="anchor" href="#strategy-1"></a>
</h3>
</div>
<div class="section level3">
<h3 id="interpretation-1">Interpretation<a class="anchor" aria-label="anchor" href="#interpretation-1"></a>
</h3>
</div>
</div>
<div class="section level2">
<h2 id="question-3">Question 3<a class="anchor" aria-label="anchor" href="#question-3"></a>
</h2>
<p>Using the PCs selected in Question 2, run the tSNE algorithm on <code>tumor</code> and identify clusters using <code>FindClusters()</code>. Plot the resulting tSNE. How many clusters are identified? Does this appear reasonable based on the structure present in the tSNE projection?</p>
<div class="section level3">
<h3 id="strategy-2">Strategy<a class="anchor" aria-label="anchor" href="#strategy-2"></a>
</h3>
</div>
<div class="section level3">
<h3 id="interpretation-2">Interpretation<a class="anchor" aria-label="anchor" href="#interpretation-2"></a>
</h3>
</div>
</div>
<div class="section level2">
<h2 id="question-4">Question 4<a class="anchor" aria-label="anchor" href="#question-4"></a>
</h2>
<p>Included in the package is a matrix called <code>tx_rates</code> containing transcription rates for ~7000 genes. Transcription rates were measured every 15 minutes for 3 hours after exposure of dendritic cells to LPS, which activates innate immunity responses. The processed data was obtained from <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE56977" class="external-link">GEO</a></p>
<p>For this exercise perform the following steps:</p>
<ol style="list-style-type: decimal">
<li><p>Exclude genes that have rates of 0 across all samples</p></li>
<li><p>Log transform the data (add a pseudocount of 1 to avoid Infinite values)</p></li>
<li><p>Normalize the data to the 0 hour time point (i.e. subtract the “rate at 0” from all of the columns )</p></li>
<li><p>run kmeans clustering on the log transformed and normalized matrix. Use k = between 5 and 10 clusters.</p></li>
<li><p>Make a heatmap with rows split based on the clusters identified in step 4. This may take a few minutes to plot, depending on your computers resources. Use the code below to generate a color palette. Discuss any interesting patterns in your clusters in the interpretation.</p></li>
</ol>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>cols_to_use <span class="ot">&lt;-</span> circlize<span class="sc">::</span><span class="fu">colorRamp2</span>(<span class="fu">c</span>(<span class="sc">-</span><span class="dv">4</span>, <span class="dv">0</span>, <span class="dv">4</span>), </span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>                                    <span class="fu">c</span>(<span class="st">"blue"</span>, <span class="st">"white"</span>, <span class="st">"red"</span>))</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="co"># pseudocode</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="fu">Heatmap</span>(mat, <span class="at">col =</span> cols_to_use, ...</span></code></pre></div>
<div class="section level3">
<h3 id="strategy-3">Strategy<a class="anchor" aria-label="anchor" href="#strategy-3"></a>
</h3>
</div>
<div class="section level3">
<h3 id="interpretation-3">Interpretation<a class="anchor" aria-label="anchor" href="#interpretation-3"></a>
</h3>
</div>
</div>
<div class="section level2">
<h2 id="bonus-super-fun-but-optional-exercise">Bonus super fun but optional exercise:<a class="anchor" aria-label="anchor" href="#bonus-super-fun-but-optional-exercise"></a>
</h2>
<p>Using your tidyverse skills plot the average transcription rate for each cluster identified in Q4 using ggplot. Facet by cluster. (Hint gather and tidy the matrix, then join with the cluster assignments)</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>An R package developed by the <a href="https://medschool.cuanschutz.edu/rbi" class="external-link">RNA Bioscience Initiative</a></p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
